File: android/keyboard/src/main/kotlin/com/viettelex/keyboard/AdjacentKeyFixer.kt
package com.viettelex.keyboard

import java.text.Normalizer
import kotlin.math.abs

/**
 * Gợi ý sửa lỗi CHẠM TRƯỢT sang phím kề (port iOS). Không tự thay — chỉ đưa lên bar.
 * Pure: engine + lexicon tiêm qua lambda.
 */
object AdjacentKeyFixer {
    private val rows: List<Pair<String, Double>> =
        listOf("qwertyuiop" to 0.0, "asdfghjkl" to 0.5, "zxcvbnm" to 1.5)

    /** Phím kề: cùng hàng ±1, hàng trên/dưới tâm lệch ≤ 0.6 phím (lưới QWERTY iOS). */
    val neighbors: Map<Char, List<Char>> = run {
        val pos = HashMap<Char, Pair<Int, Double>>()
        rows.forEachIndexed { r, (keys, off) -> keys.forEachIndexed { i, k -> pos[k] = r to off + i } }
        val out = HashMap<Char, List<Char>>()
        for ((k, p) in pos) {
            out[k] = pos.filter { (o, q) ->
                o != k && ((q.first == p.first && abs(q.second - p.second) <= 1.01) ||
                    (abs(q.first - p.first) == 1 && abs(q.second - p.second) <= 0.6))
            }.keys.sorted()
        }
        out
    }

    private fun isAsciiLetter(c: Char) = c in 'a'..'z' || c in 'A'..'Z'

    fun correction(
        raw: String,
        compose: (String) -> String,
        frequency: (String) -> Int?,
        hasCompletion: (String) -> Boolean,
    ): String? {
        val keys = raw.toCharArray()
        if (keys.size !in 2..10 || !keys.all(::isAsciiLetter)) return null
        val lower = CharArray(keys.size) { keys[it].lowercaseChar() }
        val current = compose(String(lower))
        if (frequency(current) != null || hasCompletion(current)) return null

        var bestWord: String? = null
        var bestFreq = 0
        fun consider(c: CharArray) {
            val w = compose(String(c))
            if (w == current) return
            val f = frequency(w) ?: return
            if (bestWord == null || f > bestFreq) { bestWord = w; bestFreq = f }
        }
        // Cắt tỉa theo TIỀN TỐ CHẾT (bỏ dấu thanh) — xem iOS.
        val alive = HashMap<String, Boolean>()
        fun isAlive(c: CharArray, k: Int): Boolean {
            val key = String(c, 0, k + 1)
            alive[key]?.let { return it }
            val v = hasCompletion(stripTones(compose(key)))
            alive[key] = v
            return v
        }
        fun firstDead(c: CharArray, from: Int): Int {
            var k = from
            while (k < c.size && isAlive(c, k)) k++
            return k
        }
        val dead0 = firstDead(lower, 0)
        val editable = lower.indices.filter { it <= dead0 }
        // 1 sửa: đảo 2 phím liền nhau xét trước và thắng nếu có.
        for (i in editable) {
            if (i + 1 < lower.size && lower[i] != lower[i + 1]) {
                val c = lower.copyOf(); val t = c[i]; c[i] = c[i + 1]; c[i + 1] = t; consider(c)
            }
        }
        if (bestWord == null) {
            for (i in editable) for (n in neighbors[lower[i]].orEmpty()) {
                val c = lower.copyOf(); c[i] = n; consider(c)
            }
        }
        if (bestWord == null && lower.size <= 8) {
            for (i in editable) for (a in neighbors[lower[i]].orEmpty()) {
                val ci = lower.copyOf(); ci[i] = a
                val deadI = firstDead(ci, i)
                if (deadI <= i) continue
                for (j in lower.indices) {
                    if (j <= i || j > deadI) continue
                    for (b in neighbors[lower[j]].orEmpty()) {
                        val c = ci.copyOf(); c[j] = b; consider(c)
                    }
                }
            }
        }
        var w = bestWord ?: return null
        if (keys[0].isUpperCase() && w.isNotEmpty()) w = Cp.capitalizeFirst(w)
        return w
    }

    private val toneMarks = intArrayOf(0x300, 0x301, 0x303, 0x309, 0x323)

    /** Bỏ 5 dấu thanh, giữ dấu chữ: "cũm" → "cum". */
    fun stripTones(s: String): String {
        val d = Normalizer.normalize(s, Normalizer.Form.NFD)
        val sb = StringBuilder(d.length)
        for (ch in d) if (ch.code !in toneMarks) sb.append(ch)
        return Normalizer.normalize(sb, Normalizer.Form.NFC)
    }

    /** Cache theo raw, thread-safe (gọi từ thread gợi ý nền). Đầy thì xoá sạch. */
    class Cache(private val capacity: Int = 256) {
        private val map = HashMap<String, String?>()
        fun value(raw: String, compute: () -> String?): String? {
            synchronized(map) { if (map.containsKey(raw)) return map[raw] }
            val v = compute()
            synchronized(map) {
                if (map.size >= capacity) map.clear()
                map[raw] = v
            }
            return v
        }
        val count: Int get() = synchronized(map) { map.size }
    }

    /** Bản sửa cho từ đang gõ theo setting của [bridge], qua cache. An toàn ngoài main. */
    fun lexiconCorrection(raw: String, bridge: EngineBridge): String? =
        bridge.adjacentFixCache.value(raw) {
            correction(raw,
                compose = { bridge.composeTrial(it) },
                frequency = { VNSuggest.frequency(it) },
                hasCompletion = { VNSuggest.matches(it, poolLimit = 1).isNotEmpty() })
        }
}

File: Keyboard/AdjacentKeyFixer.swift
// AdjacentKeyFixer — gợi ý sửa lỗi CHẠM TRƯỢT sang phím kề (25/09/2026; mặc định BẬT,
// tắt ở app → Kiểu Gõ).
//
// Gõ nhanh trên iPhone hay trượt sang phím bên cạnh: "nbjeeuf" (nhiều: h→b, i→j),
// "ohims" (phím: p→o), "nayd" (này: f→d), "cahcs" (cách: đảo 2 phím). Khi từ đang gõ
// KHÔNG khớp từ nào trong lexicon (kể cả làm tiền tố), thử thay phím bằng phím kề
// (1 phím, rồi 2 phím) và đảo 2 phím liền nhau; chọn phương án ÍT sửa nhất, rồi
// tần suất cao nhất. Không tự thay — chỉ đưa lên thanh gợi ý (maintainer chốt).
// Pure: engine + lexicon được tiêm qua closure — pinned by AdjacentKeyFixerTests.
import Foundation

enum AdjacentKeyFixer {
    /// QWERTY iPhone: hàng 2 thụt 0.5 phím, hàng 3 (sau shift) thụt 1.5 phím.
    private static let rows: [(keys: [Character], offset: Double)] = [
        (Array("qwertyuiop"), 0), (Array("asdfghjkl"), 0.5), (Array("zxcvbnm"), 1.5),
    ]

    /// Phím kề: cùng hàng ±1, hàng trên/dưới có tâm lệch ≤ 0.6 phím.
    static let neighbors: [Character: [Character]] = {
        var pos: [Character: (row: Int, x: Double)] = [:]
        for (r, row) in rows.enumerated() {
            for (i, k) in row.keys.enumerated() { pos[k] = (r, row.offset + Double(i)) }
        }
        var out: [Character: [Character]] = [:]
        for (k, p) in pos {
            out[k] = pos.filter { other, q in
                other != k && ((q.row == p.row && abs(q.x - p.x) <= 1.01)
                    || (abs(q.row - p.row) == 1 && abs(q.x - p.x) <= 0.6))
            }.map(\.key).sorted()
        }
        return out
    }()

    /// - raw: phím đã gõ của từ hiện tại (chữ ascii).
    /// - compose: raw → dạng hiển thị engine sẽ ra (cùng setting với bàn phím).
    /// - frequency: tần suất trong lexicon, nil = không phải từ.
    /// - hasCompletion: dạng hiển thị còn là TIỀN TỐ của từ nào đó (đang gõ dở) → không sửa.
    static func correction(raw: String,
                           compose: (String) -> String,
                           frequency: (String) -> Int?,
                           hasCompletion: (String) -> Bool) -> String? {
        let keys = Array(raw)
        guard (2...10).contains(keys.count),
              keys.allSatisfy({ $0.isASCII && $0.isLetter }) else { return nil }
        let lower = keys.map { Character($0.lowercased()) }
        let current = compose(String(lower))
        if frequency(current) != nil || hasCompletion(current) { return nil }

        var best: (word: String, freq: Int)?
        func consider(_ cand: [Character]) {
            let w = compose(String(cand))
            guard w != current, let f = frequency(w) else { return }
            if best == nil || f > best!.freq { best = (w, f) }
        }
        // Cắt tỉa theo TIỀN TỐ CHẾT: tiền tố raw[0...k] (bỏ dấu THANH — phím thanh
        // gõ sau có thể đổi/xoá thanh, "cuxmj" → cụm) không còn là tiền tố của từ
        // nào thì phím gõ thêm chỉ thu hẹp thêm → mọi chỗ sửa phải ≤ chỗ chết đầu tiên. Trước đây thử mù ~n²·25
        // lần compose (5–11 ms/phím với từ tiếng Anh dài như "keyboard"); giờ phần
        // lớn nhánh chết sau 1–2 phím. Đối chiếu brute-force trên 1500 biến thể
        // chạm trượt: kết quả trùng 100% (25/09/2026).
        var alive: [String: Bool] = [:]
        func isAlive(_ p: ArraySlice<Character>) -> Bool {
            let key = String(p)
            if let v = alive[key] { return v }
            let v = hasCompletion(stripTones(compose(key)))
            alive[key] = v
            return v
        }
        /// Chỉ số đầu tiên k ≥ from mà c[0...k] chết; c.count nếu sống hết.
        func firstDead(_ c: [Character], from: Int) -> Int {
            var k = from
            while k < c.count, isAlive(c[...k]) { k += 1 }
            return k
        }
        let dead0 = firstDead(lower, from: 0)
        let editable = lower.indices.filter { $0 <= dead0 }
        // 1 sửa. Đảo 2 phím liền nhau XÉT TRƯỚC và thắng nếu có: nó giữ nguyên đúng
        // bộ phím đã bấm — bằng chứng mạnh hơn thay phím ("cahcs" → cách, không phải
        // "các" dù "các" phổ biến hơn).
        for i in editable where i + 1 < lower.count && lower[i] != lower[i + 1] {
            var c = lower; c.swapAt(i, i + 1); consider(c)
        }
        if best == nil {
            for i in editable {
                for n in neighbors[lower[i]] ?? [] { var c = lower; c[i] = n; consider(c) }
            }
        }
        // 2 sửa (chỉ khi 1 sửa không ra): i ≤ chỗ chết của raw gốc, j ≤ chỗ chết
        // sau khi đã sửa i.
        if best == nil, lower.count <= 8 {
            for i in editable {
                for a in neighbors[lower[i]] ?? [] {
                    var ci = lower; ci[i] = a
                    let deadI = firstDead(ci, from: i)
                    guard deadI > i else { continue }        // sửa i vẫn chết ngay tại i
                    for j in lower.indices where j > i && j <= deadI {
                        for b in neighbors[lower[j]] ?? [] {
                            var c = ci; c[j] = b; consider(c)
                        }
                    }
                }
            }
        }
        guard var w = best?.word else { return nil }
        if keys.first?.isUppercase == true, let f = w.first {       // giữ hoa đầu câu
            w = f.uppercased() + w.dropFirst()
        }
        return w
    }

    /// Bỏ 5 dấu thanh, giữ dấu chữ (ư, â, đ…): "cũm" → "cum". VNSuggest coi thanh
    /// trống là tương thích mọi thanh.
    static func stripTones(_ s: String) -> String {
        let tones: Set<UInt32> = [0x300, 0x301, 0x303, 0x309, 0x323]
        var out = String.UnicodeScalarView()
        for u in s.decomposedStringWithCanonicalMapping.unicodeScalars where !tones.contains(u.value) {
            out.append(u)
        }
        return String(out).precomposedStringWithCanonicalMapping
    }

    /// Bộ nhớ đệm theo PHÍM THÔ (raw) của từ — cùng raw thì cùng kết quả (setting
    /// cố định theo EngineBridge; bridge mới = cache mới). Gõ ⌫ rồi gõ lại, hay
    /// refresh bar (textDidChange, bật/tắt bar) không tính lại. Thread-safe: bàn
    /// phím gọi từ hàng đợi gợi ý nền.
    final class Cache: @unchecked Sendable {
        private let lock = NSLock()
        private var map: [String: String?] = [:]
        private let capacity: Int
        init(capacity: Int = 256) { self.capacity = capacity }

        func value(for raw: String, compute: () -> String?) -> String? {
            lock.lock()
            if let hit = map[raw] { lock.unlock(); return hit }
            lock.unlock()
            let v = compute()
            lock.lock()
            if map.count >= capacity { map.removeAll(keepingCapacity: true) }
            map[raw] = .some(v)
            lock.unlock()
            return v
        }
        var count: Int { lock.lock(); defer { lock.unlock() }; return map.count }
    }

    /// Bản sửa cho từ đang gõ theo đúng setting của `bridge`, qua cache của bridge.
    /// An toàn gọi ngoài main: chỉ dùng composeTrial (engine scratch riêng) + lexicon tĩnh.
    static func lexiconCorrection(raw: String, bridge: EngineBridge) -> String? {
        bridge.adjacentFixCache.value(for: raw) {
            correction(raw: raw,
                       compose: { bridge.composeTrial($0) },
                       frequency: { VNSuggest.frequency(of: $0) },
                       hasCompletion: { !VNSuggest.matches($0, poolLimit: 1).isEmpty })
        }
    }
}
